Sources/AWSSDKSwiftCore/Credential/CredentialProvider.swift

```diff
@@ -26,3 +26,85 @@ extension CredentialProvider {
         return
     }
 }
+
+public struct CredentialProviderFactory {
+    
+    public struct Context {
+        let httpClient: AWSHTTPClient
+        let eventLoop: EventLoop
+    }
+    
+    private let cb: (Context) -> CredentialProvider
+    
+    private init(cb: @escaping (Context) -> CredentialProvider) {
+        self.cb = cb
+    }
+    
+    internal func createProvider(context: Context) -> CredentialProvider {
+        self.cb(context)
+    }
+}
+
+extension CredentialProviderFactory {
+    
+    public static func custom(_ factory: @escaping (Context) -> CredentialProvider) -> CredentialProviderFactory {
+        Self(cb: factory)
+    }
+    
+    public static var environment: CredentialProviderFactory {
+        Self() { _ in
+            if let accessKeyId = Environment["AWS_ACCESS_KEY_ID"],
+                let secretAccessKey = Environment["AWS_SECRET_ACCESS_KEY"] {
+                return StaticCredential(
+                    accessKeyId: accessKeyId,
+                    secretAccessKey: secretAccessKey,
+                    sessionToken: Environment["AWS_SESSION_TOKEN"]
+                )
+            } else {
+                return NullCredentialProvider()
+            }
+        }
+    }
+    
+    public static func `static`(accessKeyId: String, secretAccessKey: String, sessionToken: String? = nil) -> CredentialProviderFactory {
+        Self() { _ in
+            StaticCredential(
+                accessKeyId: accessKeyId,
+                secretAccessKey: secretAccessKey,
+                sessionToken: sessionToken
+            )
+        }
+    }
+
+    public static var ecs: CredentialProviderFactory {
+        Self() { context in
+            if let client = ECSMetaDataClient(httpClient: context.httpClient) {
+                return RotatingCredentialProvider(eventLoop: context.eventLoop, client: client)
+            }
+            
+            // fallback
+            return NullCredentialProvider()
+        }
+    }
+    
+    public static var ec2: CredentialProviderFactory {
+        Self() { context in
+            let client = InstanceMetaDataClient(httpClient: context.httpClient)
+            return RotatingCredentialProvider(eventLoop: context.eventLoop, client: client)
+        }
+    }
+    
+    public static var runtime: CredentialProviderFactory {
+        Self() { context in
+            RuntimeCredentialProvider.createProvider(
+                on: context.eventLoop,
+                httpClient: context.httpClient)
+        }
+    }
+    
```

Comment:
and the custom CredentialProviderFactory version?

Reply:
oh and the config file/shared version. I believe this'll need some work. I had to create a new provider for this. Can leave for another PR.

Reply:
oh and the custom one is in line 55.